android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/Interactor.kt

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (C) 2017. Uber Technologies
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.uber.rib.core
+
+import androidx.annotation.CallSuper
+import androidx.annotation.VisibleForTesting
+import com.jakewharton.rxrelay2.BehaviorRelay
+import com.uber.autodispose.lifecycle.CorrespondingEventsFunction
+import com.uber.autodispose.lifecycle.LifecycleEndedException
+import com.uber.autodispose.lifecycle.LifecycleScopeProvider
+import com.uber.autodispose.lifecycle.LifecycleScopes
+import com.uber.rib.core.lifecycle.InteractorEvent
+import io.reactivex.CompletableSource
+import io.reactivex.Observable
+import javax.inject.Inject
+
+/**
+ * The base implementation for all [Interactor]s.
+ *
+ * @param <P> the type of [Presenter].
+ * @param <R> the type of [Router].
+ */
+abstract class Interactor<P : Any, R : Router<*>> : LifecycleScopeProvider<InteractorEvent> {
+  @Inject
+  lateinit var injectedPresenter: P
```

Comment:
Dagger can't inject to a nullable type, but the presenter may also be passed in via constructor injection so this is split out into 2 fields